src/tools/chdman.cpp

```diff
@@ -2591,66 +2551,151 @@ static void do_extract_cd(parameters_map &params)
+	std::string trackbin_name;
 	try
 	{
-		int mode = MODE_NORMAL;
-
-		if (output_file_str->second->find(".cue") != -1)
-		{
-			mode = MODE_CUEBIN;
-		}
-		else if (output_file_str->second->find(".gdi") != -1)
-		{
-			mode = MODE_GDI;
-		}
-
-		if (cdrom->is_gdrom() && (mode == MODE_CUEBIN))
-		{
-			util::stream_format(std::cout, "Warning: extracting GD-ROM CHDs as bin/cue is not fully supported and will result in an unusable CD-ROM cue file.\n");
-		}
-
 		// process output file
 		std::error_condition filerr = util::core_file::open(*output_file_str->second, OPEN_FLAG_WRITE | OPEN_FLAG_CREATE | OPEN_FLAG_NO_BOM, output_toc_file);
 		if (filerr)
 			report_error(1, "Unable to open file (%s): %s", *output_file_str->second, filerr.message());
 
-		// process output BIN file
-		if (mode != MODE_GDI)
-		{
-			filerr = util::core_file::open(*output_bin_file_str, OPEN_FLAG_WRITE | OPEN_FLAG_CREATE, output_bin_file);
-			if (filerr)
-				report_error(1, "Unable to open file (%s): %s", *output_bin_file_str, filerr.message());
-		}
-
-		// determine total frames
 		uint64_t total_bytes = 0;
 		for (int tracknum = 0; tracknum < toc.numtrks; tracknum++)
+		{
+			// determine total frames
 			total_bytes += toc.tracks[tracknum].frames * (toc.tracks[tracknum].datasize + toc.tracks[tracknum].subsize);
 
+			// generate output filename for each track
+			std::string new_trackbin_name = *output_bin_file_str;
+
+			if (mode == MODE_GDI && toc.tracks[tracknum].trktype == cdrom_file::CD_TRACK_AUDIO)
+				new_trackbin_name += ".raw";
+			else
+				new_trackbin_name += output_bin_file_ext;
+
+			// variable replacement in output filename
+			const std::regex variables_regex("(%*)(%([+-]?\\d+)?([a-zA-Z]))");
+			std::string::const_iterator new_trackbin_name_itr = new_trackbin_name.begin();
+			std::string::const_iterator new_trackbin_name_end = new_trackbin_name.end();
+			std::string filename_formatted = new_trackbin_name;
+			std::smatch variable_matches;
+
+			while (std::regex_search(new_trackbin_name_itr, new_trackbin_name_end, variable_matches, variables_regex))
+			{
+				// full_match will always have one leading %, so if leading_escape has an even number of %s then
+				// we can know that we're working on an unescaped %
+				const std::string leading_escape = variable_matches[1].str();
+				const std::string full_match = variable_matches[2].str();
+				const std::string format_part = variable_matches[3].str();
+				const std::string format_type = variable_matches[4].str();
+
+				if ((leading_escape.size() % 2) == 0)
+				{
+					std::string replacement;
+
+					if (format_type == "t")
+					{
+						// track number
+						replacement = util::string_format("%" + format_part + "d", tracknum+1);
+
+						if (!is_splitbin)
+						{
+							util::stream_format(std::cout, "Warning: --%s not specified but track format string detected, enabling automatically\n", OPTION_OUTPUT_SPLIT);
+							is_splitbin = true;
+						}
```

Comment:
I wouldn’t do this – I’d just treat it as a literal `%t` if split binary isn’t enabled.

Also, are you checking to ensure there’s at least one `%t` if split binary *is* enabled?  If you don’t, it will overwrite the same file repeatedly, which isn’t very useful.

Reply:
>If you don’t, it will overwrite the same file repeatedly, which isn’t very useful.

Yeah that shouldn't happen, I forgot to test that specific case. I changed the code a little bit to be more saner in that regard (https://github.com/mamedev/mame/pull/12191/commits/8335ada0d17a8af3df5dcb1d3ae1681c007fc12c). Now without a track specifier in the filename it'll act the same as chdman currently does instead.

Reply:
Ok, the last few commits should address this now.

```
> ./chdman extractcd -i test.chd  -o test_split.cue -f -ob "test.bin" -sb
chdman - MAME Compressed Hunks of Data (CHD) manager 0.264 (mame0252-4086-g578558cd79a)
Input CHD:    test.chd
Output TOC:   test_split.cue
A track number variable (%t) must be specified in the output bin filename when --split is enabled

Fatal error occurred: 1
```


```
> ./chdman extractcd -i test.chd  -o test_split.cue -f -ob "test %03t.bin" -sb
chdman - MAME Compressed Hunks of Data (CHD) manager 0.264 (mame0252-4086-g578558cd79a)
Input CHD:    test.chd
Output TOC:   test_split.cue
Output Data:  test 001.bin
Output Data:  test 002.bin
Output Data:  test 003.bin
Output Data:  test 004.bin
Output Data:  test 005.bin
Output Data:  test 006.bin
Output Data:  test 007.bin
Output Data:  test 008.bin
Output Data:  test 009.bin
Output Data:  test 010.bin
Output Data:  test 011.bin
Output Data:  test 012.bin
Extraction complete
```

```
> ./chdman extractcd -i test.chd  -o test.cue -f -ob "test %03t.bin"
chdman - MAME Compressed Hunks of Data (CHD) manager 0.264 (mame0252-4086-g578558cd79a)
Input CHD:    test.chd
Output TOC:   test.cue
Output Data:  test %03t.bin
Extraction complete
```